megatron/core/inference/batch_dimensions_utils.py

```diff
@@ -0,0 +1,306 @@
+        2. Non-negative counts: Both prefill_req_count and decode_req_count must be >= 0
+        3. Token sufficiency: token_count >= prefill_req_count + decode_req_count
+
+        Sorting Rules for Attention Metadata Construction:
+        1. Batch dimensions are sorted by prefill token count (token_count - decode_req_count)
+           in descending order
+
+        Args:
+            tp_size: Tensor parallel size
+            num_cuda_graphs: Number of CUDA graphs to generate
+            cuda_graph_max_tokens: Maximum tokens for CUDA graphs
+            cuda_graph_mixed_prefill_count: Number of mixed prefill requests for CUDA graphs
+            max_requests: Maximum number of requests
+            max_tokens: Maximum total tokens
+            max_sequence_length: Maximum sequence length
+            use_cuda_graphs_for_non_decode_steps: Whether to use CUDA graphs for non-decode steps
+
+        Returns:
+            Tuple containing:
+            - List of InferenceBatchDimensions objects,
+              sorted by prefill token count in descending order
+            - Optional list of CUDA graph token counts
+        """
+
+        def add_if_valid(token_count: int, prefill_req_count: int, decode_req_count: int) -> None:
+            """Helper to create and append batch dimension to list only if it's valid."""
+            batch_dim = InferenceBatchDimensions(token_count, prefill_req_count, decode_req_count)
+            if batch_dim.is_valid(max_requests, max_sequence_length):
+                cuda_graph_batch_dimensions_list.append(batch_dim)
+
+        # Cuda graph token-counts
+        # (i.e., token counts used by cuda-graph steps, both decode and non-decode).
+        cuda_graph_token_counts = None
+        if num_cuda_graphs is not None:
+
+            # Ensure valid num_cuda_graphs.
+            if (
+                cuda_graph_max_tokens is None
+                or cuda_graph_max_tokens > max_tokens
+                or cuda_graph_max_tokens < 0
+            ):
+                cuda_graph_max_tokens = max_tokens
+            num_cuda_graphs = min(max(num_cuda_graphs, 1), cuda_graph_max_tokens)
+
+            # Cuda graph step size.
+            cuda_graph_step_size = cuda_graph_max_tokens / num_cuda_graphs
+            cuda_graph_step_size = CUDAGraphBatchDimensionBuilder.CUDA_GRAPH_ROUNDER * int(
+                math.ceil(
+                    int(cuda_graph_step_size) / CUDAGraphBatchDimensionBuilder.CUDA_GRAPH_ROUNDER
+                )
+            )
+            # Make sure divisible by TP size
+            cuda_graph_step_size = math.ceil(cuda_graph_step_size / tp_size) * tp_size
+
+            # Cuda graph token counts.
+            if num_cuda_graphs == 1:
+                cuda_graph_token_counts = [cuda_graph_max_tokens]
+            else:
+                cuda_graph_token_counts = list(
+                    range(cuda_graph_step_size, cuda_graph_max_tokens, cuda_graph_step_size)
```

Comment:
Unlikely to ever happen, but if `tp_size` > `cuda_graph_max_tokens` you will get an empty list here and an IndexError on the next line.